Add documentation on using TypeScript with data masking (#12165)
apollographql/apollo-client: docs/source/data/fragments.mdx

}
```

### Using with TypeScript

> Apollo Client provides robust TypeScript support for data masking. We've integrated data masking with [GraphQL Codegen](https://the-guild.dev/graphql/codegen) and the type format generated by its [fragment masking](https://the-guild.dev/graphql/codegen/plugins/presets/preset-client#fragment-masking) feature.
> 

Reviewer: They seem to capitalize Fragment Masking in their docs, so we should mirror it if we're referencing their feature specifically.

```suggestion
Apollo Client provides robust TypeScript support for data masking. We've integrated data masking with [GraphQL Codegen](https://the-guild.dev/graphql/codegen) and the type format generated by its [Fragment Masking](https://the-guild.dev/graphql/codegen/plugins/presets/preset-client#fragment-masking) feature.

```

> Masked types don't include fields in your operation type that are part of a fragment spread. As an example, let's use the following query.

Reviewer: I'm not sure what the clause `that are part of a fragment spread` refers to. Assuming it's the `fields` earlier in the sentence, this might be a little clearer:

```suggestion
Masked types dont't include fields from fragment spreads. As an example, let's use the following query.
```

Reviewer: Separately, is it that masked types can't include these fields, or is it that they actively remove/omit these fields?

```suggestion
Masked types omit fields in your operation type that are part of a fragment spread. As an example, let's use the following query.
```

Author: Masked types are generated without the fields from fragment spreads, so its as if those fields were never requested. I think in that regard "don't include" probably makes most sense.

```graphql
query GetCurrentUser {
  currentUser {
    id
    ...ProfileFragment
  }
}

fragment ProfileFragment on User {
  name
  age
}
```

The type definition for the query might resemble the following:

```ts
type GetCurrentUserQuery = {
  currentUser: {
    __typename: "User";
    id: string;
    name: string;
    age: number;
  } | null
}
```

<Note>
This example does not use GraphQL Codegen's true type output since it includes additional types that map scalar values differently.
</Note>

> We consider this type the _unmasked_ type since it includes fields from the `ProfileFragment` in the `GetCurrentUserQuery` type.

Reviewer: ```suggestion
This version of the `GetCurrentUserQuery` type is _unmasked_ since it includes fields from the `ProfileFragment`.
```

> Masked types on the other hand don't include fields defined in fragments.

Reviewer: ```suggestion
On the other hand, _masked_ types don't include fields defined in fragments.
```

```ts
type GetCurrentUserQuery = {
  currentUser: {
    __typename: "User";
    id: string;
    // omitted: additional internal metadata
  } | null
}
```

#### Generating masked types

> You generate masked types with either the [`typescript-operations` plugin](https://the-guild.dev/graphql/codegen/plugins/typescript/typescript-operations) or the [client preset](https://the-guild.dev/graphql/codegen/plugins/presets/preset-client). You can choose to generate masked types at any stage in the adoption process, even before you've enabled `dataMasking` in your client instance. Until you [opt in to use masked types](#opting-in-to-use-masked-types), the client will unwrap them and provide the full operation type.

Reviewer: Just cutting some words for conciseness.

```suggestion
You generate masked types with either the [`typescript-operations` plugin](https://the-guild.dev/graphql/codegen/plugins/typescript/typescript-operations) or the [client preset](https://the-guild.dev/graphql/codegen/plugins/presets/preset-client). You can generate masked types at any stage in the adoption process, even before enabling `dataMasking` in your client instance. Until you [opt in to use masked types](#opting-in-to-use-masked-types), the client unwraps them and provides the full operation type.
```

> The following sections provide detail on how to configure GraphQL Codegen to output masked types for you.

Reviewer: ```suggestion
The following sections show how to configure GraphQL Codegen to output masked types.
```

##### With the `typescript-operations` plugin

<Note>
Support for the `@unmask` directive was introduced with `@graphql-codegen/typescript-operations` [v4.4.0](https://github.com/dotansimha/graphql-code-generator/releases/tag/release-1732308151614)
</Note>

Add the following configuration to your GraphQL Codegen config.

```ts title="codegen.ts"
const config: CodegenConfig = {
  // ...
  generates: {
    "path/to/types.ts": {
      plugins: ["typescript-operations"],
      config: {
        // ...
        inlineFragmentTypes: "mask",
        customDirectives: {
          apolloUnmask: true
        }
      }
    }
  }
}
```

##### With the `client-preset`

<Note>
Support for the `@unmask` directive was introduced with `@graphql-codegen/client-preset` [v4.5.1](https://github.com/dotansimha/graphql-code-generator/releases/tag/release-1732308151614)
</Note>

Add the following configuration to your GraphQL Codegen config.

> <Caution>
> The [fragment masking](https://the-guild.dev/graphql/codegen/plugins/presets/preset-client#fragment-masking) feature in the client preset is incompatible with Apollo Client's data masking feature. You need to [disable fragment masking](https://the-guild.dev/graphql/codegen/plugins/presets/preset-client#how-to-disable-fragment-masking) in your configuration (included below). If you use the [generated `useFragment` function](https://the-guild.dev/graphql/codegen/plugins/presets/preset-client#the-usefragment-helper), you will need to use Apollo Client's [`useFragment`](#usefragment) hook instead.
> </Caution>

Reviewer: We generally try to avoid the term `disabled` [for inclusivity].(https://developers.google.com/style/word-list#disable)

```suggestion
<Caution>
The [Fragment Masking](https://the-guild.dev/graphql/codegen/plugins/presets/preset-client#fragment-masking) feature in the client preset is incompatible with Apollo Client's data masking feature. You need to [turn off Fragment Masking](https://the-guild.dev/graphql/codegen/plugins/presets/preset-client#how-to-disable-fragment-masking) in your configuration (included below). If you use the [generated `useFragment` function](https://the-guild.dev/graphql/codegen/plugins/presets/preset-client#the-usefragment-helper), you must use Apollo Client's [`useFragment`](#usefragment) hook instead.
</Caution>
```

```ts title="codegen.ts"
const config: CodegenConfig = {
  // ...
  generates: {
    "path/to/gql/": {
      preset: "client",
      presetConfig: {
        // ...
        // disables the incompatible GraphQL Codegen fragment masking feature
        fragmentMasking: false,
        inlineFragmentTypes: "mask",
        customDirectives: {
          apolloUnmask: true
        }
      }
    }
  }
}
```

#### Opting in to use masked types

<Note>
We recommend that you opt in to use masked types only after you've [enabled `dataMasking`](#enabling-data-masking) in your `ApolloClient` instance.
</Note>

> By default, the client will unwrap operation types and provide the full operation result type. To prevent the client from unwrapping these types and use the masked type directly, you need to opt in to using them. This strategy allows for [incremental adoption](#incremental-adoption-in-an-existing-application) in your application and avoids the need to update large parts of your application to satisfy the change in types.

Reviewer: ```suggestion
By default, the client unwraps operation types and provides the full operation result type. To prevent this behavior and have the client use masked types, you need to opt in. This strategy allows for [incremental adoption](#incremental-adoption-in-an-existing-application) in your application and avoids the need to update large parts of your application to satisfy the change in types.
```

You can opt in to use the masked types in one of two ways.

##### Opting in globally

> To opt in to using masked types throughout your entire application at once, you modify the `DataMasking` exported type from `@apollo/client` using TypeScript's [declaration merging](https://www.typescriptlang.org/docs/handbook/declaration-merging.html) ability.

Reviewer: Just more conciseness suggestions:

```suggestion
To turn on masked types for your whole application at once, modify the `DataMasking` exported type from `@apollo/client` using TypeScript's [declaration merging](https://www.typescriptlang.org/docs/handbook/declaration-merging.html) ability.
```

<Tip>
We recommend this approach for most cases. Use this approach with the [`@unmask` directive](./directives#unmask) for the best path to [incremental adoption](#incremental-adoption-in-an-existing-application).
</Tip>

> First create a TypeScript file that will be used to modify the `DataMasking` type.

Reviewer: ```suggestion
First, create a TypeScript file that will be used to modify the `DataMasking` type.
```

```ts title="apollo-client.d.ts"
// This import is necessary to ensure all Apollo Client imports
// are still available to the rest of the application.
import '@apollo/client';

declare module "@apollo/client" {
  interface DataMasking {
    enabled: true;
  }
}
```

<Note>

> This example uses `apollo-client.d.ts` as the file name to make it easily identifiable, but you can name this file however you wish.

Reviewer: ```suggestion
This example uses `apollo-client.d.ts` as the file name to make it easily identifiable. You can name this file as you wish.
```

</Note>

With `enabled` set to `true`, any request-based API will type `data` using the masked type and prevent the client from unwrapping it.

##### Opting in per operation

> If you prefer an incremental approach, you can opt in to use masked types per operation. This can be useful when your application creates multiple Apollo Client instances where only a subset of them enable data masking.

Reviewer: ```suggestion
If you prefer an incremental approach, you can opt in to use masked types per operation. This can be useful when your application creates multiple Apollo Client instances where only a subset enables data masking.
```

Apollo Client provides a `Masked` helper type that tells the client to use the masked type directly. You can use this with `TypedDocumentNode` or with generic arguments.

```ts
import { Masked, TypedDocumentNode } from "@apollo/client";

// With TypedDocumentNode
const QUERY: TypedDocumentNode<Masked<QueryType>, VarsType> = gql`
  # ...
`;

// with generic arguments
const { data } = useQuery<Masked<QueryType>, VarsType>(QUERY)
```

The use of `TypedDocumentNode` with the `Masked` type is common enough that Apollo Client provides a `MaskedDocumentNode` convenience type as a replacement for `TypedDocumentNode`. It is simply a shortcut for `TypedDocumentNode<Masked<QueryType>, VarsType>`.

```ts
import { MaskedDocumentNode } from "@apollo/client";

const QUERY: MaskedDocumentNode<QueryType, VarsType> = gql`
  # ...
`;
```

#### Using with fragments

> When using [colocated fragments](#colocating-fragments) with your components, its best to ensure the object passed to your component is done in a type safe way. This means:

Reviewer: ```suggestion
When using [colocated fragments](#colocating-fragments) with your components, it's best to ensure the object passed to your component is done in a type-safe way. This means:
```

```ts
function PostDetails({ post }: PostDetailsProps) {
  // ...

  post.title
  // ❌ Property 'title' does not exist on type '{ " $fragmentRefs"?: { PostDetailsFragment: PostDetailsFragment; } | undefined; }'
}
```

`FragmentType` also prevents parent components from accidentally omitting fragment spreads for child components, regardless of whether the field selection satisfies the fragment's data requirements.

```tsx title="Posts.tsx"
const GET_POSTS = gql`
  query GetPosts {
    posts {
      id
      title
      shortDescription
    }
  }
`;

export default function Posts() {
  // ...

  return (
    <div>
      {allPosts.map((post) => (
        <PostDetails key={post.id} post={post} />
        // ❌ Type '{ __typename: "Post"; id: string; title: string; shortDescription: string; }' has no properties in common with type '{ " $fragmentRefs"?: { PostDetailsFragment: PostDetailsFragment; } | undefined; }'.
      ))}
    </div>
  );
}
```

In this example, the `GetPosts` query selects enough fields to satisfy the `PostDetails` data requirements, but TypeScript warns us because the `PostDetailsFragment` was not included in the `GetPosts` query.

#### Unwrapping masked types

> On rare occasion you may need access to the unmasked type of a particular operation. Apollo Client provides the `Unmasked` helper type that unwraps masked types and removes meta information on the type.

Reviewer: ```suggestion
On rare occasions, you may need access to the unmasked type of a particular operation. Apollo Client provides the `Unmasked` helper type that unwraps masked types and removes meta information on the type.
```

<Note>

> This is the same helper type used by the client when unwrapping types while data masking is disabled or for APIs that use the full result.

Reviewer: ```suggestion
This is the same helper type the client uses when unwrapping types while data masking is turned off or for APIs that use the full result.
```

Masked types don't include fields in your operation type that are part of a fragment spread. As an example, let's use the following query.

```graphql
query GetCurrentUser {
  currentUser {
    id
    ...ProfileFragment
  }
}

fragment ProfileFragment on User {
  name
  age
}
```

The type definition for the query might resemble the following:

```ts
type GetCurrentUserQuery = {
  currentUser: {
    __typename: "User";
    id: string;
    name: string;
    age: number;
  } | null
}
```

<Note>
This example does not use GraphQL Codegen's true type output since it includes additional types that map scalar values differently.
</Note>

We consider this type the _unmasked_ type since it includes fields from the `ProfileFragment` in the `GetCurrentUserQuery` type.

Masked types on the other hand don't include fields defined in fragments.

```ts
type GetCurrentUserQuery = {
  currentUser: {
    __typename: "User";
    id: string;
  } | null
}
```

<Note>
GraphQL Codegen generates masked types with additional meta information on the type that is omitted in this example.

> </Note>

Author: I like that better. I think I overused notes, so this is a good tradeoff while still communicating that information. Great idea!

The following sections provide detail on how to configure GraphQL Codegen to output masked types for you.

##### With the `typescript-operations` plugin

<Note>
We require [v4.4.0](https://github.com/dotansimha/graphql-code-generator/releases/tag/release-1732308151614) as the minimum supported version of `@graphql-codegen/typescript-operations` which includes support for the `@unmask` directive.
</Note>

Add the following configuration to your GraphQL Codegen config.

```ts title="codegen.ts"
const config: CodegenConfig = {
  // ...
  generates: {
    "path/to/types.ts": {
      plugins: ["typescript-operations"],
      config: {
        // ...
        inlineFragmentTypes: "mask",
        customDirectives: {
          apolloUnmask: true
        }
      }
    }
  }
}

> ````

Author: oops good catch

##### With the `client-preset`

<Note>

> We require [v4.5.1](https://github.com/dotansimha/graphql-code-generator/releases/tag/release-1732308151614) as the minimum supported version of `@graphql-codegen/client-preset` which includes support for the `@unmask` directive.

Author: I dig it!
